libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,672 @@
+            param_saved[pname] = get_param_value(param_s(pname));
+            BIT_CLEAR(param_changed, pname);
+        }
+    }
+}
+
+AP_Quicktune::param_s AP_Quicktune::get_pname(AP_Quicktune::axis_names axis, AP_Quicktune::stages stage)
+{
+    switch (axis)
+    {
+        case axis_names::RLL:
+            switch (stage)
+            {
+                case stages::P:
+                    return param_s::RLL_P;
+                case stages::D:
+                    return param_s::RLL_D;
+                case stages::FLTT:
+                    return param_s::RLL_FLTT;
+                case stages::FLTD:
+                    return param_s::RLL_FLTD;
+                default:
+                    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_pname - axis was %d, stage was %d", uint8_t(axis), uint8_t(stage));
+                    return param_s::END;
+            }
+        case axis_names::PIT:
+            switch (stage)
+            {
+                case stages::P:
+                    return param_s::PIT_P;
+                case stages::D:
+                    return param_s::PIT_D;
+                case stages::FLTT:
+                    return param_s::PIT_FLTT;
+                case stages::FLTD:
+                    return param_s::PIT_FLTD;
+                default:
+                    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_pname - axis was %d, stage was %d", uint8_t(axis), uint8_t(stage));
+                    return param_s::END;
+            }
+        case axis_names::YAW:
+            switch (stage)
+            {
+                case stages::P:
+                    return param_s::YAW_P;
+                case stages::D:
+                    return param_s::YAW_D;
+                case stages::FLTT:
+                    return param_s::YAW_FLTT;
+                case stages::FLTD:
+                    return param_s::YAW_FLTD;
+                default:
+                    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_pname - axis was %d, stage was %d", uint8_t(axis), uint8_t(stage));
+                    return param_s::END;
+            }
+        default:
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_pname - axis was %d", uint8_t(axis));
+            // INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return param_s::END;
+    }
+}
+
+AP_Quicktune::stages AP_Quicktune::get_stage(AP_Quicktune::param_s param)
+{
+    if (param == param_s::RLL_P || param == param_s::PIT_P || param == param_s::YAW_P){
+        return stages::SMAX;
+    } else if (param == param_s::RLL_I || param == param_s::PIT_I || param == param_s::YAW_I){
+        return stages::I;
+    } else if (param == param_s::RLL_FF || param == param_s::PIT_FF || param == param_s::YAW_FF){
+        return stages::FF;
+    } else {
+        return stages::END;
+    }
+}
+
+AP_Float *AP_Quicktune::get_param_pointer(AP_Quicktune::param_s param)
+{
+    switch (param)
```

Comment:
I the PID objects are the same, so you could switch on axis to get the object and then switch on P/I/D/SMax/ect

Reply:
done